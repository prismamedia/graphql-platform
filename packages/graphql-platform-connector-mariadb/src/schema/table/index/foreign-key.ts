import * as core from '@prismamedia/graphql-platform';
import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/memoize';
import { escapeIdentifier } from '../../../escaping.js';
import type { MariaDBConnector } from '../../../index.js';
import type { Table } from '../../table.js';
import { AbstractIndex } from '../abstract-index.js';
import { ReferenceColumn } from '../column/reference.js';

export interface ForeignKeyIndexConfig {
  /**
   * Optional, the index's name
   */
  name?: utils.Nillable<string>;
}

/**
 * @see https://mariadb.com/kb/en/foreign-keys/
 */
export class ForeignKeyIndex extends AbstractIndex {
  public readonly config?: ForeignKeyIndexConfig;
  public readonly configPath: utils.Path;

  public constructor(
    table: Table,
    public readonly edge: core.Edge<any, MariaDBConnector>,
  ) {
    super(table);

    // config
    {
      this.config = edge.config.foreignKey;
      this.configPath = utils.addPath(edge.configPath, 'foreignKey');

      utils.assertNillablePlainObjectConfig(this.config, this.configPath);
    }
  }

  @Memoize()
  public get referencedTable(): Table {
    return this.table.schema.getTableByNode(this.edge.head);
  }

  @Memoize()
  public get columns(): ReadonlyArray<ReferenceColumn> {
    return this.table.getColumnTreeByEdge(this.edge).columns;
  }

  @Memoize()
  public get name(): string {
    const nameConfig = this.config?.name;
    const nameConfigPath = utils.addPath(this.configPath, 'name');

    if (nameConfig) {
      if (typeof nameConfig !== 'string') {
        throw new utils.UnexpectedConfigError('a string', nameConfig, {
          path: nameConfigPath,
        });
      }

      if (nameConfig.length > 64) {
        throw new utils.UnexpectedConfigError(
          'an identifier shorter than 64 characters',
          nameConfig,
          { path: nameConfigPath },
        );
      }

      return nameConfig;
    }

    return this.table.schema.namingStrategy.getForeignKeyIndexName(
      this.table.name,
      this.edge,
      this.columns,
    );
  }

  /**
   * @see https://mariadb.com/kb/en/create-table/#foreign-key
   */
  @Memoize()
  public get definition(): string {
    return [
      'FOREIGN KEY',
      escapeIdentifier(this.name),
      `(${this.columns.map(({ name }) => escapeIdentifier(name)).join(',')})`,
      'REFERENCES',
      escapeIdentifier(this.referencedTable.name),
      `(${this.columns
        .map(({ referencedColumn: { name } }) => escapeIdentifier(name))
        .join(',')})`,
      'ON UPDATE RESTRICT',
      'ON DELETE RESTRICT',
    ]
      .filter(Boolean)
      .join(' ');
  }
}
