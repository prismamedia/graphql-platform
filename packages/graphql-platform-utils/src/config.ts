export * from './config/name.js';
export * from './config/optional-deprecation.js';
export * from './config/optional-description.js';
export * from './config/optional-flag.js';
