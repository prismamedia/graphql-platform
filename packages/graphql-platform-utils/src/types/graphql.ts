export {
  FieldConfig as GraphQLFieldConfig,
  FieldConfigMap as GraphQLFieldConfigMap,
  FieldResolver as GraphQLFieldResolver,
  FieldResolverArgs as GraphQLFieldResolverArgs,
} from './graphql/field';
export { OperationType as GraphQLOperationType } from './graphql/operation-type';
