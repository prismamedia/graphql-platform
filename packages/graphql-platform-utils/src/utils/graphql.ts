export * from './graphql/get-enum-type';
export * from './graphql/list-decorator';
export * from './graphql/non-null-decorator';
export * from './graphql/parse-resolver-info';
