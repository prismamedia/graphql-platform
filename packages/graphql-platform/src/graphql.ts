import type * as graphql from 'graphql';
import type { ConnectorInterface } from './connector-interface.js';
import type { GraphQLPlatform } from './index.js';

export type GPBoundGraphQLFieldResolver<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TSource,
  TArgs = any,
  TResult = unknown,
> = (
  this: GraphQLPlatform<TRequestContext, TConnector>,
  source: TSource,
  args: TArgs,
  context: TRequestContext,
  info: graphql.GraphQLResolveInfo,
) => TResult;

export interface GPBoundGraphQLFieldConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TSource,
  TArgs = any,
  TResult = unknown,
> extends graphql.GraphQLFieldConfig<TSource, TRequestContext, TArgs> {
  resolve?: GPBoundGraphQLFieldResolver<
    TRequestContext,
    TConnector,
    TSource,
    TArgs,
    TResult
  >;
}
