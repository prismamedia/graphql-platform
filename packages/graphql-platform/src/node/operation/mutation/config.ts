import type * as utils from '@prismamedia/graphql-platform-utils';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { CreationConfig } from './abstract-creation.js';
import type { DeletionConfig } from './abstract-deletion.js';
import type { UpdateConfig } from './abstract-update.js';

export type {
  CreationConfig,
  PostCreateArgs,
  PreCreateArgs,
} from './abstract-creation.js';
export type {
  DeletionConfig,
  PostDeleteArgs,
  PreDeleteArgs,
} from './abstract-deletion.js';
export type {
  PostUpdateArgs,
  PreUpdateArgs,
  UpdateConfig,
} from './abstract-update.js';

export interface MutationConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TContainer extends object,
> {
  [utils.MutationType.CREATION]: CreationConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.UPDATE]: UpdateConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.DELETION]: DeletionConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
}
