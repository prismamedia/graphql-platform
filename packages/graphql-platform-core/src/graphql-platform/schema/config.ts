import { GraphQLOperationType } from '@prismamedia/graphql-platform-utils';
import { GraphQLFieldConfigMap, GraphQLObjectType, GraphQLSchemaConfig as BaseGraphQLSchemaConfig } from 'graphql';
import { OperationType } from '../operation';
import { ResourceMap } from '../resource/map';

export interface GraphQLSchemaConfigConfig
  extends Omit<BaseGraphQLSchemaConfig, 'mutation' | 'query' | 'subscription'> {
  resourceMap: ResourceMap;
  queryMap: GraphQLFieldConfigMap<any, any>;
  mutationMap: GraphQLFieldConfigMap<any, any>;
}

export class GraphQLSchemaConfig implements BaseGraphQLSchemaConfig {
  public constructor(readonly config: GraphQLSchemaConfigConfig) {}

  public getOperationTypeFieldMap<TType extends OperationType>(type: TType): GraphQLFieldConfigMap<any, any> {
    const fields: GraphQLFieldConfigMap<any, any> = {};

    for (const resource of this.config.resourceMap.values()) {
      for (const operation of resource.getOperations(type)) {
        if (operation.isPublic()) {
          fields[operation.name] = operation.getGraphQLFieldConfig();
        }
      }
    }

    return fields;
  }

  public get mutation() {
    return new GraphQLObjectType({
      name: 'Mutation',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Mutation), ...this.config.mutationMap },
    });
  }

  public get query() {
    return new GraphQLObjectType({
      name: 'Query',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Query), ...this.config.queryMap },
    });
  }

  public get subscription() {
    return undefined;
  }
}
