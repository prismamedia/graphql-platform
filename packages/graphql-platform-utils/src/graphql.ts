import * as graphql from 'graphql';
import { getOptionalFlag, OptionalFlag } from './config.js';
import { getEnumKeys } from './enum.js';
import { UnexpectedValueError } from './error.js';
import { indefinite } from './indefinite.js';
import { isNil, type Nillable } from './nil.js';
import type { Path } from './path.js';
import { isPlainObject, type PlainObject } from './plain-object.js';

export const isGraphQLResolveInfo = (
  maybeGraphQLResolveInfo: unknown,
): maybeGraphQLResolveInfo is graphql.GraphQLResolveInfo =>
  isPlainObject(maybeGraphQLResolveInfo) &&
  typeof maybeGraphQLResolveInfo['fieldName'] === 'string' &&
  Array.isArray(maybeGraphQLResolveInfo['fieldNodes']) &&
  graphql.isOutputType(maybeGraphQLResolveInfo['returnType']) &&
  graphql.isOutputType(maybeGraphQLResolveInfo['parentType']);

export function assertGraphQLResolveInfo(
  maybeGraphQLResolveInfo: unknown,
  path?: Path,
): asserts maybeGraphQLResolveInfo is graphql.GraphQLResolveInfo {
  if (!isGraphQLResolveInfo(maybeGraphQLResolveInfo)) {
    throw new UnexpectedValueError(
      `a GraphQLResolveInfo`,
      maybeGraphQLResolveInfo,
      { path },
    );
  }
}

export const isGraphQLASTNode = <TKind extends graphql.Kind>(
  maybeGraphQLASTNode: unknown,
  kind: TKind,
): maybeGraphQLASTNode is graphql.ASTKindToNode[TKind] =>
  isPlainObject(maybeGraphQLASTNode) && maybeGraphQLASTNode['kind'] === kind;

export function assertGraphQLASTNode<TKind extends graphql.Kind>(
  maybeGraphQLASTNode: unknown,
  kind: TKind,
  path?: Path,
): asserts maybeGraphQLASTNode is graphql.ASTKindToNode[TKind] {
  if (!isGraphQLASTNode(maybeGraphQLASTNode, kind)) {
    throw new UnexpectedValueError(`a GraphQL ${kind}`, maybeGraphQLASTNode, {
      path,
    });
  }
}

export const parseGraphQLScalarValue = <TInternal>(
  type: graphql.GraphQLScalarType<TInternal, any>,
  maybeScalarValue: unknown,
  path?: Path,
): Nillable<TInternal> => {
  if (isNil(maybeScalarValue)) {
    return maybeScalarValue;
  }

  try {
    return type.parseValue(maybeScalarValue);
  } catch (error) {
    throw new UnexpectedValueError(indefinite(type.name), maybeScalarValue, {
      cause: error,
      path,
    });
  }
};

export const parseGraphQLEnumValue = <TInternal = any>(
  type: graphql.GraphQLEnumType,
  maybeEnumValue: unknown,
  path?: Path,
): Nillable<TInternal> => {
  if (isNil(maybeEnumValue)) {
    return maybeEnumValue;
  }

  const enumValue = type
    .getValues()
    .find(({ value }) => value === maybeEnumValue);

  if (!enumValue) {
    throw new UnexpectedValueError(
      `${indefinite(type.name)} (= a value among "${type
        .getValues()
        .map(({ value }) => value)
        .join(', ')}")`,
      maybeEnumValue,
      { path },
    );
  }

  return enumValue.value;
};

export const parseGraphQLLeafValue = (
  type: graphql.GraphQLLeafType,
  maybeLeafValue: unknown,
  path?: Path,
): any => {
  if (isNil(maybeLeafValue)) {
    return maybeLeafValue;
  }

  return graphql.isScalarType(type)
    ? parseGraphQLScalarValue(type, maybeLeafValue, path)
    : parseGraphQLEnumValue(type, maybeLeafValue, path);
};

export type CreateGraphQLEnumTypeOptions = {
  description?: Nillable<string>;
  useKeyAsValue?: OptionalFlag;
};

export const createGraphQLEnumType = (
  name: string,
  enumerable: PlainObject,
  options?: CreateGraphQLEnumTypeOptions,
): graphql.GraphQLEnumType =>
  new graphql.GraphQLEnumType({
    name,
    description: options?.description,
    values: Object.fromEntries(
      getEnumKeys(enumerable).map<[string, graphql.GraphQLEnumValueConfig]>(
        (key) => [
          key,
          {
            value: getOptionalFlag(options?.useKeyAsValue, false)
              ? undefined
              : enumerable[key],
          },
        ],
      ),
    ),
  });
