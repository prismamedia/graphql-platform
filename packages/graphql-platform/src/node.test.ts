import {
  isPlainObject,
  MutationType,
  mutationTypes,
  type Name,
} from '@prismamedia/graphql-platform-utils';
import * as graphql from 'graphql';
import _ from 'lodash';
import { GraphQLPlatform } from './index.js';
import {
  ComponentConfig,
  Edge,
  NodeConfig,
  NodeName,
  NodeSelection,
  OnHeadDeletion,
  ReverseEdgeMultiple,
} from './node.js';
import { Article, Category, nodes } from './__tests__/config.js';

const hasEnabledMutationByType = (
  gp: GraphQLPlatform,
  type?: MutationType,
): boolean =>
  [...gp.operationsByNameByType.mutation.values()].some(
    (mutation) =>
      mutation.isEnabled() && (!type || mutation.mutationTypes.includes(type)),
  );

const hasPublicMutationByType = (
  gp: GraphQLPlatform,
  type?: MutationType,
): boolean =>
  [...gp.operationsByNameByType.mutation.values()].some(
    (mutation) =>
      mutation.isPublic() && (!type || mutation.mutationTypes.includes(type)),
  );

describe('Node', () => {
  describe('Fails', () => {
    describe('Invalid name', () => {
      it.each([
        [
          '-InvalidName',
          `\"GraphQLPlatformConfig.nodes.-InvalidName\" - Expects to be valid against the GraphQL \"Names\" specification (@see: https://spec.graphql.org/draft/#sec-Names), got: '-InvalidName'
└ Cause: Names must start with [_a-zA-Z] but \"-InvalidName\" does not.`,
        ],
        [
          'invalidName',
          `\"GraphQLPlatformConfig.nodes.invalidName\" - Expects to be in PascalCase (= \"InvalidName\"), got: 'invalidName'`,
        ],
      ])(
        `throws an Error on invalid name: %s`,
        (invalidName, expectedError) => {
          expect(
            () =>
              new GraphQLPlatform({
                // @ts-expect-error
                nodes: { [invalidName]: {} },
              }),
          ).toThrowError(expectedError);
        },
      );
    });

    describe('Invalid component(s)', () => {
      it.each([
        [
          {},
          `"GraphQLPlatformConfig.nodes.Test.components" - Expects at least one "component", got: undefined`,
        ],
        [
          { components: undefined },
          `"GraphQLPlatformConfig.nodes.Test.components" - Expects at least one "component", got: undefined`,
        ],
        [
          { components: null },
          `"GraphQLPlatformConfig.nodes.Test.components" - Expects at least one "component", got: null`,
        ],
        [
          { components: {} },
          `"GraphQLPlatformConfig.nodes.Test.components" - Expects at least one "component", got: {}`,
        ],
      ])(
        `throws an Error on invalid components: %p`,
        (config, expectedError) => {
          expect(
            () =>
              new GraphQLPlatform({
                nodes: {
                  // @ts-expect-error
                  Test: config,
                },
              }),
          ).toThrowError(expectedError);
        },
      );

      describe('Invalid leaf(s)', () => {});

      describe('Invalid edge(s)', () => {
        it('throws an Error on missing "head"', () => {
          expect(
            () =>
              new GraphQLPlatform({
                nodes: {
                  ...nodes,
                  Article: {
                    ...Article,
                    components: {
                      ...Article.components,
                      edgeToAMissingModel: {
                        kind: 'Edge',
                        head: 'MissingModel',
                      },
                    },
                  },
                },
              }),
          ).toThrowError(
            '"GraphQLPlatformConfig.nodes.Article.components.edgeToAMissingModel.head" - Expects a "node"\'s name among "Article, Category, Tag, ArticleTag, User, UserProfile, Log", got: \'MissingModel\'',
          );
        });

        it('throws an Error on missing "head reference"', () => {
          expect(
            () =>
              new GraphQLPlatform({
                nodes: {
                  ...nodes,
                  Article: {
                    ...Article,
                    components: {
                      ...Article.components,
                      edgeToAMissingReference: {
                        kind: 'Edge',
                        head: 'Article.missingUnique',
                      },
                    },
                  },
                },
              }),
          ).toThrowError(
            '"GraphQLPlatformConfig.nodes.Article.components.edgeToAMissingReference.head" - Expects a "unique-constraint"\'s name among "_id, id, category_slug", got: \'missingUnique\'',
          );
        });

        it('throws an Error on referencing itself', () => {
          expect(
            () =>
              new GraphQLPlatform({
                nodes: {
                  Category: {
                    ...Category,
                    components: {
                      ...Category.components,
                      parent: {
                        kind: 'Edge',
                        head: 'Category.parent_slug',
                      },
                    },
                  },
                },
              }),
          ).toThrowError(
            '"GraphQLPlatformConfig.nodes.Category.components.parent.head" - Expects a "unique-constraint" not refering itself, got: \'parent_slug\'',
          );
        });

        it('throws an Error on invalid "onHeadDeletion"', () => {
          expect(
            () =>
              new GraphQLPlatform({
                nodes: {
                  Test: {
                    mutation: false,

                    components: {
                      id: {
                        kind: 'Leaf',
                        type: 'UUIDv4',
                        nullable: false,
                        mutable: false,
                      },
                      parent: {
                        kind: 'Edge',
                        head: 'Test',
                        onHeadDeletion: OnHeadDeletion.CASCADE,
                      },
                      brother: {
                        kind: 'Edge',
                        head: 'Test',
                        onHeadDeletion: OnHeadDeletion.SET_NULL,
                      },
                    },

                    uniques: [['id']],
                  },
                },
              }),
          )
            .toThrowError(`"GraphQLPlatformConfig.nodes.Test.components" - 2 errors:
└ "parent.onHeadDeletion" - Expects not to be "CASCADE" as the node "Test" cannot be deleted, got: 'CASCADE'
└ "brother.onHeadDeletion" - Expects not to be "SET_NULL" as the edge "Test.brother" is immutable, got: 'SET_NULL'`);
        });
      });
    });

    describe('Invalid unique-constraint(s)', () => {
      it(`throws an Error on empty config`, () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                Test: {
                  components: {
                    _id: { kind: 'Leaf', type: 'UnsignedInt' },
                  },
                  uniques: [],
                },
              },
            }),
        ).toThrowError(
          `"GraphQLPlatformConfig.nodes.Test.uniques" - Expects at least one "unique-constraint", got: []`,
        );
      });

      it(`throws an Error on nullable identifier`, () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                Test: {
                  components: {
                    _id: { kind: 'Leaf', type: 'UnsignedInt' },
                  },
                  uniques: [['_id']],
                },
              },
            }),
        ).toThrowError(
          '"GraphQLPlatformConfig.nodes.Test.uniques.0" - Expects its identifier (= the first unique constraint, composed of the component "_id") to be non-nullable (= at least one of its components being non-nullable)',
        );
      });

      it(`throws an Error on mutable identifier`, () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                Test: {
                  components: {
                    _id: { kind: 'Leaf', type: 'UnsignedInt', nullable: false },
                  },
                  uniques: [['_id']],
                },
              },
            }),
        ).toThrowError(
          '"GraphQLPlatformConfig.nodes.Test.uniques.0" - Expects its identifier (= the first unique constraint, composed of the component "_id") to be immutable (= all its components being immutable)',
        );
      });

      it('throws an Error on empty components', () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                ...nodes,
                Article: {
                  ...Article,
                  uniques: [[]],
                },
              },
            }),
        ).toThrowError(
          '"GraphQLPlatformConfig.nodes.Article.uniques.0" - Expects at least one "component", got: []',
        );
      });

      it('throws an Error on unknown component', () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                ...nodes,
                Article: {
                  ...Article,
                  uniques: [['missingComponent']],
                },
              },
            }),
        ).toThrowError(
          `\"GraphQLPlatformConfig.nodes.Article.uniques.0.0\" - Expects a \"component\"'s name among \"_id, id, status, title, slug, body, category, createdBy, createdAt, updatedBy, updatedAt, metas, highlighted, sponsored\", got: 'missingComponent'`,
        );
      });
    });

    describe('Invalid reverse-edge(s)', () => {
      it('throws an Error on provided configuration when no referrers have been found', () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                ...nodes,
                Log: {
                  ...nodes.Log,
                  reverseEdges: {
                    // @ts-expect-error
                    myExtraReverseEdge: {},
                  },
                },
              },
            }),
        ).toThrowError(
          `\"GraphQLPlatformConfig.nodes.Log.reverseEdges\" - Expects no configuration as there is no node having an edge heading to the \"Log\" node, got: { myExtraReverseEdge: {} }`,
        );
      });

      it('throws an Error on unknown head', () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                ...nodes,
                User: {
                  ...nodes.User,
                  reverseEdges: {
                    ...nodes.User.reverseEdges,
                    invalidEdge: { originalEdge: 'UnknownModel' },
                  },
                },
              },
            }),
        ).toThrowError(
          `\"GraphQLPlatformConfig.nodes.User.reverseEdges.invalidEdge.originalEdge\" - Expects a node having an edge heading to the \"User\" node (= a value among \"Article, UserProfile\"), got: 'UnknownModel'`,
        );
      });

      it('throws an Error on extra configuration', () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                ...nodes,
                User: {
                  ...nodes.User,
                  reverseEdges: {
                    ...nodes.User.reverseEdges,
                    invalidEdge: { originalEdge: 'Article.unknownEdgeName' },
                  },
                },
              },
            }),
        ).toThrowError(
          `\"GraphQLPlatformConfig.nodes.User.reverseEdges.invalidEdge.originalEdge\" - Expects no more configuration for \"Article\"'s edge as there is no more edge heading to the \"User\" node`,
        );
      });

      it('throws an Error on unknown edge', () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                ...nodes,
                User: {
                  ...nodes.User,
                  reverseEdges: {
                    invalidEdge: { originalEdge: 'Article.unknownEdgeName' },
                  },
                },
              },
            }),
        ).toThrowError(
          `\"GraphQLPlatformConfig.nodes.User.reverseEdges.invalidEdge.originalEdge\" - Expects an edge heading to the \"User\" node (= a value among \"createdBy, updatedBy\"), got: 'unknownEdgeName'`,
        );
      });

      it('throws an Error on invalid edge', () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                ...nodes,
                User: {
                  ...nodes.User,
                  reverseEdges: {
                    invalidEdge: { originalEdge: 'Article.category' },
                  },
                },
              },
            }),
        ).toThrowError(
          `\"GraphQLPlatformConfig.nodes.User.reverseEdges.invalidEdge.originalEdge\" - Expects an edge heading to the \"User\" node (= a value among \"createdBy, updatedBy\"), got: 'category'`,
        );
      });
    });

    describe('Invalid output', () => {
      it("throws an Error on invalid virtual-field's dependsOn", () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                ...nodes,
                Article: {
                  ...nodes.Article,
                  output: {
                    ...nodes.Article.output,
                    virtualFields: {
                      myInvalidField: {
                        dependsOn: '{ unknownField }',
                        type: graphql.GraphQLString,
                      },
                    },
                  },
                },
              },
            }),
        )
          .toThrowError(`\"GraphQLPlatformConfig.nodes.Article.output.virtualFields.myInvalidField.dependsOn\" - Expects a valid fragment, got: '{ unknownField }'
└ Cause: Expects an \"Article\"'s field, got: 'unknownField'`);
      });
    });

    describe('Invalid mutation', () => {
      it('throws an Error on update-able node without update-able component', () => {
        expect(
          () =>
            new GraphQLPlatform({
              nodes: {
                Hit: {
                  components: {
                    id: {
                      kind: 'Leaf',
                      type: 'UUIDv4',
                      nullable: false,
                      mutable: false,
                    },
                  },
                  uniques: [['id']],
                },
              },
            }),
        ).toThrowError(
          `\"GraphQLPlatformConfig.nodes.Hit\" - Expects at least one mutable component as it is mutable: either disable the \"update\" or enable it on one of its components`,
        );
      });
    });
  });

  describe('Works', () => {
    const gp = new GraphQLPlatform({ nodes });

    it.each([
      [
        'Article',
        [
          '_id',
          'id',
          'status',
          'title',
          'slug',
          'body',
          'category',
          'createdBy',
          'createdAt',
          'updatedBy',
          'updatedAt',
          'metas',
          'highlighted',
          'sponsored',
        ],
        ['_id', 'id', 'category_slug'],
        ['tags'],
      ],
      [
        'Category',
        ['_id', 'id', 'title', 'slug', 'parent', 'order'],
        ['_id', 'id', 'parent_slug', 'parent_order'],
        ['children', 'articles'],
      ],
      [
        'Tag',
        ['id', 'deprecated', 'title', 'slug', 'createdAt', 'updatedAt'],
        ['id', 'slug'],
        ['articles'],
      ],
      [
        'ArticleTag',
        ['article', 'tag', 'order'],
        ['article_tag', 'article_order'],
        [],
      ],
      [
        'User',
        ['id', 'username', 'createdAt', 'lastLoggedInAt'],
        ['id', 'username'],
        ['createdArticles', 'updatedArticles', 'profile'],
      ],
      [
        'UserProfile',
        ['user', 'birthday', 'facebookId', 'googleId', 'twitterHandle'],
        ['user'],
        [],
      ],
      ['Log', ['_id', 'message', 'createdAt'], ['_id'], []],
    ])(
      '"%s" has valid definition',
      (nodeName, componentNames, uniqueConstraintNames, reverseEdgeNames) => {
        const node = gp.getNode(nodeName);

        expect(node.name).toBe(nodeName);
        expect([...node.componentsByName.keys()]).toEqual(componentNames);
        expect([...node.uniqueConstraintsByName.keys()]).toEqual(
          uniqueConstraintNames,
        );
        expect([...node.reverseEdgesByName.keys()]).toEqual(reverseEdgeNames);
        expect(node.selection).toBeInstanceOf(NodeSelection);
      },
    );

    it.each([
      ['Article', ['_id'], ['_id'], []],
      ['Category', ['_id'], ['_id'], []],
      ['Tag', [], [], []],
      ['ArticleTag', [], [], []],
      ['User', ['createdAt', 'lastLoggedInAt'], [], []],
      ['UserProfile', [], [], []],
    ])(
      '"%s" is public but has private component(s) / unique-constraint(s) / reverse-edge(s)',
      (
        nodeName,
        privateComponentNames,
        privateUniqueConstraintNames,
        privateReverseEdgeNames,
      ) => {
        const node = gp.getNode(nodeName);

        expect(node.isPublic()).toBeTruthy();

        expect(
          [...node.componentsByName.values()]
            .filter((component) => !component.isPublic())
            .map((component) => component.name),
        ).toEqual(privateComponentNames);

        expect(
          [...node.uniqueConstraintsByName.values()]
            .filter((uniqueConstraint) => !uniqueConstraint.isPublic())
            .map((uniqueConstraint) => uniqueConstraint.name),
        ).toEqual(privateUniqueConstraintNames);

        expect(
          [...node.reverseEdgesByName.values()]
            .filter((reverseEdge) => !reverseEdge.isPublic())
            .map((reverseEdge) => reverseEdge.name),
        ).toEqual(privateReverseEdgeNames);
      },
    );

    it.each(['Log'])(
      '"%s" is private so cannot have public component(s) / unique-constraint(s) / reverse-edge(s)',
      (nodeName) => {
        const node = gp.getNode(nodeName);

        expect(node.isPublic()).toBeFalsy();

        expect(
          [...node.componentsByName.values()].some((component) =>
            component.isPublic(),
          ),
        ).toBeFalsy();

        expect(
          [...node.uniqueConstraintsByName.values()].some((uniqueConstraint) =>
            uniqueConstraint.isPublic(),
          ),
        ).toBeFalsy();

        expect(
          [...node.reverseEdgesByName.values()].some((reverseEdge) =>
            reverseEdge.isPublic(),
          ),
        ).toBeFalsy();
      },
    );

    it(`can have (reverse-)edge heading to itself`, () => {
      const Category = gp.getNode('Category');

      const parent = Category.getEdge('parent');
      expect(parent).toBeInstanceOf(Edge);
      expect(parent.head).toBe(Category);
      expect(parent.referencedUniqueConstraint.name).toEqual('_id');

      const children = Category.getReverseEdge('children');
      expect(children).toBeInstanceOf(ReverseEdgeMultiple);
      expect(children.head).toBe(Category);

      expect(children.originalEdge).toBe(parent);
    });

    describe('mutation can be configured', () => {
      it('can be disabled all at once', () => {
        const gp = new GraphQLPlatform({
          nodes: Object.fromEntries(
            Object.entries(nodes).map<[NodeName, NodeConfig]>(
              ([name, config]) => [
                name,
                {
                  ...config,
                  mutation: false,
                  components: Object.fromEntries(
                    Object.entries(config.components).map<
                      [Name, ComponentConfig<any, any>]
                    >(([name, config]) => [
                      name,
                      (config.kind === 'Edge'
                        ? _.omit(config, ['onHeadDeletion'])
                        : config) as any,
                    ]),
                  ),
                },
              ],
            ),
          ),
        });

        expect(hasEnabledMutationByType(gp)).toBeFalsy();
      });
    });

    describe.each(mutationTypes)('%s can be configured', (mutationType) => {
      it.each([
        { [mutationType]: false },
        { [mutationType]: { enabled: false } },
      ])(
        'can be disabled using the following configuration: %p',
        (mutationConfig) => {
          const gp = new GraphQLPlatform({
            nodes: Object.fromEntries(
              Object.entries(nodes).map<[Name, NodeConfig]>(
                ([name, config]) => [
                  name,
                  {
                    ...config,
                    mutation: {
                      ...(isPlainObject(config.mutation)
                        ? config.mutation
                        : {}),
                      ...mutationConfig,
                    },
                    components: Object.fromEntries(
                      Object.entries(config.components).map<
                        [Name, ComponentConfig<any, any>]
                      >(([name, config]) => [
                        name,
                        (config.kind === 'Edge'
                          ? _.omit(config, ['onHeadDeletion'])
                          : config) as any,
                      ]),
                    ),
                  },
                ],
              ),
            ),
          });

          expect(hasEnabledMutationByType(gp, mutationType)).toBeFalsy();
        },
      );
    });
  });
});
