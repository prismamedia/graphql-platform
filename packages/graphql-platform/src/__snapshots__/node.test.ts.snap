// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node Article has a node: Article 1`] = `
"# The article is the main node, written by the journalists
type Article implements Node {
  # This UUID is used to identify an Article publicly
  id: UUID!
  status: ArticleStatus!
  title: NonEmptyTrimmedString!
  slug: NonEmptyTrimmedString!

  # The article's body
  body: NonEmptyTrimmedString
  category: Category

  # Either this \\"Article\\" node has a(n) \\"category\\" edge to a(n) \\"Category\\" node or not?
  hasCategory: Boolean!
  createdBy: User!
  createdAt: DateTime!
  updatedBy: User!
  updatedAt: DateTime!

  # Contains any arbitrary data you want to store alongside the article
  metas: JSONObject
}"
`;

exports[`Node ArticleTag has a node: ArticleTag 1`] = `
"type ArticleTag {
  article: Article!
  tag: Tag!
  order: NonNegativeInt!
}"
`;

exports[`Node Category has a node: Category 1`] = `
"type Category implements Node {
  id: UUID!
  title: NonEmptyTrimmedString!
  slug: NonEmptyTrimmedString!
  parent: Category

  # Either this \\"Category\\" node has a(n) \\"parent\\" edge to a(n) \\"Category\\" node or not?
  hasParent: Boolean!
  order: NonNegativeInt!
}"
`;

exports[`Node Hit has a node: Hit 1`] = `
"type Hit {
  _id: NonNegativeInt!
  at: DateTime!
}"
`;

exports[`Node Tag has a node: Tag 1`] = `
"type Tag implements Node {
  id: UUID!
  title: NonEmptyTrimmedString!
  slug: NonEmptyTrimmedString!
  createdAt: DateTime!
  updatedAt: DateTime!
}"
`;

exports[`Node User has a node: User 1`] = `
"type User implements Node {
  id: UUID!
  username: NonEmptyTrimmedString!
}"
`;

exports[`Node UserProfile has a node: UserProfile 1`] = `
"type UserProfile {
  user: User!
  birthday: Date
  facebookId: ID
  googleId: ID
  twitterId: ID
}"
`;
